pkg/state/history.go

```diff
@@ -0,0 +1,86 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package state
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/lib/pq"
+	"github.com/xataio/pgroll/pkg/migrations"
+)
+
+// Migration represents a single migration in the migration history
+// of a schema
+type Migration struct {
+	Migration migrations.Migration
+	CreatedAt time.Time
+}
+
+// SchemaHistory returns all migrations applied to a schema in ascending
+// timestamp order
+func (s *State) SchemaHistory(ctx context.Context, schema string) ([]Migration, error) {
+	rows, err := s.pgConn.QueryContext(ctx,
+		fmt.Sprintf("SELECT name, migration, created_at "+
+			"FROM %s.migrations "+
+			"WHERE schema=$1 ORDER BY created_at",
+			pq.QuoteIdentifier(s.schema)), schema)
+	if err != nil {
+		return nil, err
+	}
+
+	defer rows.Close()
+
+	var entries []Migration
+	for rows.Next() {
+		var name, rawMigration string
+		var createdAt time.Time
+
+		rows.Scan(&name, &rawMigration, &createdAt)
+		if err != nil {
+			return nil, err
+		}
+
+		var mig migrations.Migration
+		json.Unmarshal([]byte(rawMigration), &mig)
+		if err != nil {
+			return nil, err
+		}
+
+		entries = append(entries, Migration{
+			Migration: mig,
+			CreatedAt: createdAt,
+		})
+	}
+
+	return entries, nil
+}
+
+// WriteToFile writes the migration to a file in the target directory,
+// optionally prepending the filename with the migration timestamp
+func (m *Migration) WriteToFile(targetDir string, withTimestamp bool) error {
+	var prefix string
+	if withTimestamp {
+		prefix = m.CreatedAt.Format(time.RFC3339Nano) + "_"
+	}
+
+	fileName := fmt.Sprintf("%s%s.json", prefix, m.Migration.Name)
+	filePath := filepath.Join(targetDir, fileName)
+
+	bytes, err := json.MarshalIndent(m.Migration, "", "  ")
+	if err != nil {
+		return err
+	}
+
+	// Ensure the target directory exists
+	err = os.MkdirAll(targetDir, 0o755)
+	if err != nil {
+		return err
+	}
+
+	return os.WriteFile(filePath, bytes, 0o600)
```

Comment:
nit: You could create a file first and use a json encoder to write directly to it rather than writing to memory and then the file.

Reply:
thanks, done